# Introduksjon til REST API og requests-biblioteket

I denne notebooken skal vi lære hvordan vi kan hente data fra internett ved
hjelp av API-er (Application Programming Interface). Vi starter enkelt og bygger
opp til å hente værdata fra Yr.no sitt åpne API.

## Læringsmål
- Forstå hva et REST API er
- Lære å bruke requests-biblioteket
- Hente og behandle JSON-data
- Arbeide med Yr.no sitt vær-API
- Bygge enkle funksjoner for værhenting

## Del 1: Hva er et API?

Du kjenner gjerne til forkortelsen UI, som står for User Interface. Dette er grensesnittet mellom bruker og datamaskin. For et GUI (Graphical User Interface) er dette grensesnittet visuelt, som de fleste grensesnitt vi er vant til å bruke er. Altså er dette alt som vises på skjermen - knapper, tekst, lenker, vindu osv. Du som bruker forteller dataprogrammet hva du ønsker skal skje eller vises på skjermen ved å trykke på knapper eller bruke andre funksjoner i grensesnittet. 

Et API er et annet type grensesnitt, og står for Application Programming Interface. Som vi kan gjette fra navnet, er ikke dette grensesnittet beregnet for en bruker, men for andre dataprogrammer. Et API er altså et grensesnitt der ulike programmer kan utveksle informasjon, sende kommandoer osv, der ett program får et annet program til å gjøre noe. 

### REST API

REST API er et eksempel på en type API som er velkjent og mye brukt. Også kjent som RESTful API, er dette et API som følger arkiteturen / designprinsippene til Representational State Transfer (REST).

> Hva er arkitektur / designprinsipper ?
>
> I programmering kommer du ofte borti ord som arkitektur (architecture) og design. 
> Dette refererer til regler og prinsipper som kodere og programmer bør eller skal 
> følge. Det kan være strenge regler som må oppfylles, eller mer løse retningslinjer,
> og gjerne en kombinasjon av disse. En slik arkitektur-spesifikasjon gjør at du som
> programmerer vet hva du kan forvente når du leser andres kode, så den blir lettere
> å forstå og vedlikeholde, og du vet også hva du kan forvente av et annet program
> du skal benytte gjennom et API. Designprinsipper kan også bidra til å ivareta
> viktige hensyn som sikkerhet, effektivitet osv.

REST API er en arkitektur som er tiltenkt for å designe nettverkstjenester, særlig nettbaserte APIer. Arkitekturen tilbyr en fleksibel plattform for utviklere som også er konsistent og effektiv. Den bruker HTTP protokollen for å utveksle informasjon, og kan implementeres i alle programmeringsspråk. 

REST bygger på følgende prinsipper:

- Uniform Interface (enhetlig grensesnitt)
    - Server responderer likt på samme forespørsler, uavhengig av hvilken applikasjon som sender forespørselen.
    - Det er faste standardiserte metoder for å hente eller manipulere ressurser fra server.
- Client-Server decoupling
    - Klient og server kjører helt uavhengig av hverandre
    - Klienten kommuniserer bare med server på designerte URI-adresser
    - Server kan bare svare klienten på forespørsler, og ellers ikke samhandle med klient på andre måter
- Stateless
    - Hver forespørsel fra klient må inneholde all nødvendig informasjon for at server skal prosessere forespørselen.
    - Server lagrer ikke data om klientens tilstand. Hvis klientens tilstand er relevant for forespørselen må altså klienten selv sende slik informasjon sammen med forespørselen til server.
- Cacheable
    - Når det er mulig skal ressurser kunne lagres i cache (mellomlagres / bufres) hos klient eller server. Dette for å øke ytelse og unngå unødvendig overføring av samme data flere ganger.
    - Responser inneholder informasjon om leverte ressurser kan mellomlagres.
- Layered system architecture (lagdelt system)
    - Kommunikasjon mellom server og klient kan gå gjennom flere lag av "mellommenn", som proxyer, gatewayer eller belastningsfordelere. 
    - REST er designet slik at hverken server eller klient kan vite om de kommuniserer direkte med hverandre, eller gjennom slike lag.
- Code on Demand (valgfritt)
    - REST kan levere script og kjørbar kode til klienten, for å utvide funksjonalitet på klientsiden.
    - Dette benyttes ofte i sammenheng med nettsider og web-applikasjoner, der klienten henter JavaScript kode fra server og kjører den i nettleseren.


## REST API i praksis - HTTP forespørsler

Det vi oftest bruker REST API til er å hente eller manipulere data på en server. På serveren er gjerne denne informasjonen lagret i databaser, så det kan være greit å ha databasestrukturer i bakhodet når du tenker på hvordan du skal samhandle med et REST API. Som nevnt brukes HTTP-forespørsler i kommunikasjon med server. Under er en kjapp oversikt over de ulike typene av HTTP-forespørsler: 

- `GET`: Hent data fra server
- `POST`: Send data som server skal gjøre noe med
- `PUT`: Send data som skal erstatte eller opprette ny ressurs på server
- `PATCH`: Send data som delvis oppdaterer eksisterende ressurs på server
- `DELETE`: Slett dataressurs fra server

Mest brukt er GET, etterfulgt av POST. PUT, PATCH og DELETE er mer strukturerte forespørsler som ligner på databaseoperasjoner. Det viktigste å huske er at GET brukes for å hente data fra server, og POST brukes for å sende data til server. 

> Merk: GET requests kan også inneholde data / informasjon
> 
> Ofte trenger vi å sende en god del data sammen med en GET forespørsel, for at server skal kunne vite hvilken informasjon som skal sendes tilbake. Dette kan inkludere en token som autentiserer klienten. Altså er det ikke bare med POST vi kan sende data til server, men når vi sender data vi ønsker server skal behandle, som å legge til en ny kunde i en database, bruker vi POST. For å hente en kunde bruker vi GET.


## Del 2: Oppgaver

### Import av biblioteker

Kjør cellen under for å importere nødvendige biblioteker til å kjøre kode i andre celler senere.

In [4]:
# Importer bibliotekene vi trenger
import requests
import json
from datetime import datetime
import pandas as pd

# For automatiske tester trenger vi pytest og ipytest
import pytest
import ipytest
ipytest.autoconfig()

## Oppgave 1: Din første API-forespørsel

La oss starte med å forstå hvordan en enkel API-forespørsel fungerer. Kjør cellen under og se på output.

In [ ]:
# Lag din første API-forespørsel
response_3a = requests.get('https://httpbin.org/get')

print("Statuskode:", response_3a.status_code)
print("Responstype:", type(response_3a))
print("\nResponsinnhold:")
print(response_3a.text)

**Spørsmål til refleksjon:**
1. Hva betyr status kode 200?
2. Hva tror du skjer hvis vi prøver en URL som ikke eksisterer?

**Oppgave 1a:** Prøv å endre URL-en til noe som ikke eksisterer og se hva som skjer:

In [6]:
# Din kode her - prøv en ugyldig URL. 
# Følg kode fra forrige kodecelle, men bruk en ugyldig URL og skriv ut statuskoden.


**Oppgave 1b:** Les om HTTP-statuskoder og skriv kort om de mest vanlige statuskodene.

_Skriv inn ditt svar her_

## Oppgave 2: Forstå JSON-data

De fleste API-er returnerer data i JSON-format (JavaScript Object Notation). JSON er lett å lese både for mennesker og datamaskiner. 

Kjør cellen under og se på resultatet

In [27]:
# Hent JSON-data fra test-API
response_3a = requests.get('https://httpbin.org/json')

print("Status kode:", response_3a.status_code)

if response_3a.status_code == 200:
    # Konverter JSON-tekst til Python dictionary
    data = response_3a.json()
    
    print("\nData type:", type(data))
    print("\nJSON data (formatert):")
    print(json.dumps(data, indent=2)) # json.dumps formaterer utskriften så den blir lettere å lese
                                      # Vi kunne skrevet ut data direkte, men da hadde alt kommet på en linje
else:
    print(f"Feil! Status kode: {response_3a.status_code}")

Status kode: 503
Feil! Status kode: 503


**Oppgave 2a:** Noen raske spørsmål om koden over
- Hvilken datatype konverteres en json respons til i python?
- Hvordan henter du ut json-data fra en get-respons i python?
- Hvilken datatype har respons-variabelen over, og hva kan du si om innholdet i variabelen?

_Skriv ditt svar her_

**Oppgave 2b:** Hent ut spesifikke verdier
- Sørg for at kodecellen over kjørte ok før du går videre til å kjøre kode i cellen under.
- Bruk variabelen data fra oppgaven over til å hente ut slideshow-tittel og forfatter

In [8]:
# Variabelen data skal, dersom cellen over kjørte uten feil, 
# nå inneholde data fra responsen. Vi printer den her for å
# vise at den er tilgjengelig.
print(data)

# Skriv din kode her for å hente ut slideshow-tittel og forfatter
# Print ut verdiene med print()-funksjonen
# Du kan også hente ut andre verdier og skrive ut disse



{'slideshow': {'author': 'Yours Truly', 'date': 'date of publication', 'slides': [{'title': 'Wake up to WonderWidgets!', 'type': 'all'}, {'items': ['Why <em>WonderWidgets</em> are great', 'Who <em>buys</em> WonderWidgets'], 'title': 'Overview', 'type': 'all'}], 'title': 'Sample Slide Show'}}


**Oppgave 2c:** POST-request
- Lag noen verdier i dictionary mine_data under, og kjør cellen.

In [9]:
mine_data = {
    # Lag en dictionary med noen nøkkel-verdi par
}
response_3a = requests.post('https://httpbin.org/post', json=mine_data)
print("Status kode:", response_3a.status_code)
if response_3a.status_code == 200:
    data = response_3a.json()
    print("\nPOST respons data (formatert):")
    print(json.dumps(data, indent=2))
else:
    print("Oops, noe gikk galt med POST-forespørselen")

Status kode: 200

POST respons data (formatert):
{
  "args": {},
  "data": "{}",
  "files": {},
  "form": {},
  "headers": {
    "Accept": "*/*",
    "Accept-Encoding": "gzip, deflate",
    "Content-Length": "2",
    "Content-Type": "application/json",
    "Host": "httpbin.org",
    "User-Agent": "python-requests/2.32.5",
    "X-Amzn-Trace-Id": "Root=1-68e659b7-12ec62d005d584da0b70ed72"
  },
  "json": {},
  "origin": "185.73.27.4",
  "url": "https://httpbin.org/post"
}


**Oppgave 2c:** Hva skjedde her?
- Hvordan sender man en post-request i python?
- Hvilket argument i denne funksjonen brukes for å sende data fra en dictionary?
Se på data som ble skrevet ut fra responsen
- Hva tror du dette er?
- Kan du kjenne igjen noen verdier i responsen?

_Skriv svaret ditt her_

## Oppgave 3: Introduksjon til Yr.no API

Nå skal vi begynne å arbeide med ekte værdata! Yr.no tilbyr gratis værdata gjennom sitt API.

**API dokumentasjon:** https://api.met.no/weatherapi/locationforecast/2.0/documentation
Gå inn på lenken over og ha denne tilgjengelig mens du jobber videre. Siden inneholder API-spesifiksjonen, og forteller utviklere hvordan de kan bruke yr sin locationforecast api. 

**Base-adresse:** REST APIer bruker å ha en base-adresse, som kommer først for alle uri-er knyttet til APIet. 
For locations apiet til yr er baseadressen `https://api.met.no/weatherapi/locationforecast/2.0/`. Ofte, som her,
inkluderer base-uri et versjonsnummer. Slik kan APIet oppdateres samtidig som den gamle versjonen er tilgjengelig,
slik at utviklere har tid til å oppdatere sine programmer til å støtte spesifikasjon i ny versjon av APIet.

> Yr har flere APIer som du kan utforske på https://api.met.no
> Vi skal bruke loactionforecast som eksempel her

La oss prøve et enkelt kall til APIet, med uri `https://api.met.no/weatherapi/locationforecast/2.0/status`. 
Dette skal gi oss en status-beskjed, og kan brukes for å sjekke om vi klarer å koble oss opp mot APIet.

In [10]:
response_3a = requests.get('https://api.met.no/weatherapi/locationforecast/2.0/status')
print("Status kode:", response_3a.status_code)

Status kode: 403


Oops.. Her fikk vi en feil statuskode - 403

Hva betyr denne statuskoden?

_Skriv svaret ditt her_

Hvis vi ser i dokumentasjonen, står det i delen 
[AUTHENTICATION](https://api.met.no/weatherapi/locationforecast/2.0/documentation#AUTHENTICATION) 
at en er nødt til å bruke en unik identifikator i User-Agent header-verdi for at ikke forespørselen
skal avvises.

La oss sjekke hvordan våre header-verdier så ut fra requesten vi sendte over

> Tips:
> Informasjon om sendte forespørsler kan hentes ut fra respons-variabelen
> med egenskapen `request` (eks: `respons_variabel.request`)

In [11]:
# Her skriver vi ut headere som ble sendt med forespørselen
# Kjør cellen og se på output
print("\nRequest headers:")
request_headers = response_3a.request.headers
for key, value in request_headers.items():
    print(f"{key}: {value}")


Request headers:
User-Agent: python-requests/2.32.5
Accept-Encoding: gzip, deflate
Accept: */*
Connection: keep-alive


Her ser vi at vår User-Agent er satt til "python-requests/" + versjonsnummer. Dette oppfyller ikke vilkårene til i API-spesifikasjonen. La oss prøve igjen der vi setter en unik verdi for User-Agent:

In [12]:
# Kjør cellen under for å gjøre et nytt kall med en gyldig User-Agent header

# Sett en unik User-Agent header som beskrevet i dokumentasjonen
headers = {
    'User-Agent': 'TestApp/1.0 (it-test@sanpro.no)'
}

# Gjør forespørselen med de nye headerne
response_3a = requests.get('https://api.met.no/weatherapi/locationforecast/2.0/status', headers=headers)

print("Status kode:", response_3a.status_code)

# Skriv ut headere som ble sendt med forespørselen
print("\nRequest headers:")
print(json.dumps(dict(response_3a.request.headers), indent=2))

# Hvis statuskoden er 200 (ok), skriv ut JSON-responsen
if response_3a.status_code == 200:
    data = response_3a.json()
    print("\nRespons data:")
    print(json.dumps(data, indent=2))

Status kode: 200

Request headers:
{
  "User-Agent": "TestApp/1.0 (it-test@sanpro.no)",
  "Accept-Encoding": "gzip, deflate",
  "Accept": "*/*",
  "Connection": "keep-alive"
}

Respons data:
{
  "last_update": "2025-10-08T12:27:29Z"
}


Nå fikk vi et gyldig svar! 

- Hva inneholdt responsen av data?

_Skriv svaret ditt her_

> Tips ISO 8601
> Les om standarden ISO 8601 for dato og tid formatert som tekst på [Wikipedia](https://en.wikipedia.org/wiki/ISO_8601)

Som vi nå har erfart er det lurt å sjekke dokumentasjon på API før bruk, og også vilkår. Les yr sine [Terms of Service her](https://api.met.no/doc/TermsOfService)

**Viktige punkter om Yr.no API:**
- Du må identifisere applikasjonen din med en "User-Agent" header
- Du kan ikke sende for mange forespørsler (maks 20 per sekund)
- API-et krever geografiske koordinater (bredde- og lengdegrad)

## Et enkelt kall til locationforecast

Siden det er noen verdier vi kommer til å bruke flere ganger videre, kan det være lurt å lagre disse til noen variabler først. 

Kjør cellen under for at variablene som defineres skal være tilgjengelig senere.

In [13]:
# Sett opp korrekte headers for Yr.no API
headers = {
    'User-Agent': 'TestApp/1.0 (it-test@sanpro.no)'  # Bytt gjerne ut med din e-post
}

# URL for Yr.no API
base_url = "https://api.met.no/weatherapi/locationforecast/2.0/"
base_url_compact = base_url + "compact"
base_url_complete = base_url + "complete"

# Utskrift av base-verdiene
print("API URL:", base_url)
print("API url for compact metoden:", base_url_compact)
print("API url for complete metoden:", base_url_complete)
print("Headers:", headers)

API URL: https://api.met.no/weatherapi/locationforecast/2.0/
API url for compact metoden: https://api.met.no/weatherapi/locationforecast/2.0/compact
API url for complete metoden: https://api.met.no/weatherapi/locationforecast/2.0/complete
Headers: {'User-Agent': 'TestApp/1.0 (it-test@sanpro.no)'}


Da kan vi gjøre vårt første ordentlige kall til APIet:

In [14]:

# Koordinater for Hana, Sandnes
lat = 58.854  # Breddegrad (N-S)
long = 5.767  # Lengdegrad (Ø-V)

# Bygg URL for API-forespørsel
url = f"{base_url_compact}?lat={lat}&lon={long}"

# Gjennomfør forespørselen
response_3a = requests.get(url, headers=headers)
print("Status kode:", response_3a.status_code)
if response_3a.status_code == 200:
    print("Respons data (formatert):")
    data = response_3a.json()
    print(json.dumps(data, indent=2)[:500], "...")  # Print de første 500 tegnene av responsen


Status kode: 200
Respons data (formatert):
{
  "type": "Feature",
  "geometry": {
    "type": "Point",
    "coordinates": [
      5.767,
      58.854,
      44
    ]
  },
  "properties": {
    "meta": {
      "updated_at": "2025-10-08T12:27:12Z",
      "units": {
        "air_pressure_at_sea_level": "hPa",
        "air_temperature": "celsius",
        "cloud_area_fraction": "%",
        "precipitation_amount": "mm",
        "relative_humidity": "%",
        "wind_from_direction": "degrees",
        "wind_speed": "m/s"
      }
    },
     ...


**Oppgave 3a:** Lag din første værdata-forespørsel

Velg en lokasjon og hent værdata for denne lokasjonen med et enkelt API-kall. Skriv koden din under. 

**Lagre responsen i en variabel med navn `response_3a`**

_Tips: Bruk google maps til å finne koordinatene_

In [15]:
# Skriv din kode her
# OBS! Lagre responsen i en variabel med navn response_3a, slik at vi kan bruke den i neste oppgave.

In [16]:
# Kjør denne cellen for å sjekke at du gjorde oppgaven over riktig.

def test_response_3a():
    assert 'response_3a' in globals(), "Variabelen response_3a finnes ikke."
    assert response_3a.status_code == 200, "Responsen i variabelen response_3a har ikke statuskode 200."

ok = ipytest.run()
if ok == 0:
    print("\n", "*"*50)
    print("Oppgave 3a er gjennomført korrekt!")

.                                                                                            [100%]
1 passed in 0.04s

 **************************************************
Oppgave 3a er gjennomført korrekt!


## Oppgave 4: Utforsk værdata-strukturen

Før vi kan bruke dataene, må vi forstå hvordan de er organisert.

In [17]:
# Konverter til JSON og lagre til fil
if response_3a.status_code == 200:
    data_task4 = response_3a.json()
    with open('weather_data.json', 'w') as f:
        json.dump(data_task4, f, indent=2)
    print("Data lagret til weather_data.json")
else:
    print("Feil! Respons ugyldig.")
    print("Sjekk om du har gjennomført forrige oppgave korrekt, og kjørt cellen.")
    print("OG at du har lagret responsen i en variabel med navn response_3a.")

Data lagret til weather_data.json


Åpne filen `weather_data.json` og se gjennom verdiene. Filen er ganske stor, men du vil se at strukturen gjentas etter 50-100 linjer, så du trenger bare å forstå starten for å skjønne hele innholdet. 

- Hvilke hovedegenskaper (1. nivå) er det i json-filen? (det skal være tre stk, navngi dem)
- Hvilke egenskaper finnes under "properties" ?
- Hvor finner du de faktiske værdataene i filen?

**Oppgave 4a:** Se nærmere på strukturen til en enkelt prognose



In [18]:
def get_properties(locationforecast_compact_response=data_task4):
    """Hent ut properties fra responsen fra locationforecast compact API-et"""
    return locationforecast_compact_response['properties']

def get_timeseries(locationforecast_compact_response=data_task4):
    """Hent ut prognoser fra responsen fra locationforecast compact API-et."""
    return locationforecast_compact_response['properties']['timeseries']


In [19]:
def test_get_properties():
    properties = get_properties()
    assert isinstance(properties, dict), "get_properties skal returnere en dictionary."
    assert 'timeseries' in properties, "properties skal inneholde nøkkelen 'timeseries'."
    assert isinstance(properties['timeseries'], list), "properties['timeseries'] skal være en liste."
    
def test_get_timeseries():
    timeseries = get_timeseries()
    assert isinstance(timeseries, list), "get_timeseries skal returnere en liste."
    assert len(timeseries) > 0, "get_timeseries skal returnere en ikke-tom liste."
    first_entry = timeseries[0]
    assert 'time' in first_entry, "Hver entry i timeseries skal inneholde nøkkelen 'time'."
    assert 'data' in first_entry, "Hver entry i timeseries skal inneholde nøkkelen 'data'."
    

ok = ipytest.run()
if ok == 0:
    print("\n", "*"*50)
    print("Oppgave 4a er gjennomført korrekt!")

...                                                                                          [100%]
3 passed in 0.03s

 **************************************************
Oppgave 4a er gjennomført korrekt!


In [20]:
# Utforsk en enkelt prognose i detalj
def print_timeseries_keys():
    """Skriv ut alle nøklene i en enkelt prognose i timeseries-listen"""
    prognoser = get_timeseries()
    first_item = prognoser[0]
    def print_keys(d, prefix=''):
        for key, value in d.items():
            print(f"{prefix}{key}")
            if isinstance(value, dict):
                print_keys(value, prefix + '  ')
    print_keys(first_item)
    
print_timeseries_keys()

time
data
  instant
    details
      air_pressure_at_sea_level
      air_temperature
      cloud_area_fraction
      relative_humidity
      wind_from_direction
      wind_speed
  next_12_hours
    summary
      symbol_code
    details
  next_1_hours
    summary
      symbol_code
    details
      precipitation_amount
  next_6_hours
    summary
      symbol_code
    details
      precipitation_amount


Forklar utskriften over.
- Hvilken informasjon finnes i ett timeseries objekt?

## Oppgave 5: Lag en klasse for værdata

Nå som vi forstår datastrukturen, kan vi enklere hente ut data vi ønsker.

**Oppgave 5a:** Lag en klasse, Prognosis, som 

In [21]:
def get_weather_now(lat, lon):
    """
    Henter værprognose for nåværende tidspunkt, på gitte koordinater.
    
    Args:
        lat (float): Breddegrad
        lon (float): Lengdegrad
    
    Returns:
        dict: Værdata eller None hvis feil oppstod
    """
if oslo_weather:
    print("🌤️ Nåværende vær i Oslo:")
    print(f"Tidspunkt: {oslo_weather['tidspunkt']}")
    print(f"Temperatur: {oslo_weather['temperatur_celsius']}°C")
    print(f"Luftfuktighet: {oslo_weather['luftfuktighet_prosent']}%")
    print(f"Vindhastighet: {oslo_weather['vindhastighet_ms']} m/s")
else:
    print("❌ Kunne ikke hente værdata")

NameError: name 'oslo_weather' is not defined

**Oppgave 5a:** Test funksjonen med koordinater for din hjemby

In [ ]:
# Test med andre koordinater
# Bergen: 60.3913, 5.3221
# Trondheim: 63.4305, 10.3951
# Tromsø: 69.6492, 18.9553

# Din kode her:
# my_weather = get_current_weather(LAT, LON)
# print(my_weather)

## Oppgave 6: Hent værprognose for flere timer

API-et gir oss prognoser for mange timer fremover. La oss lage en funksjon som henter prognoser for de neste timene.

In [ ]:
def get_hourly_forecast(lat, lon, hours=12):
    """
    Henter timesprognose for de neste X timene.
    
    Args:
        lat (float): Breddegrad
        lon (float): Lengdegrad
        hours (int): Antall timer å hente prognose for
    
    Returns:
        list: Liste med værprognoser
    """
    
    headers = {'User-Agent': 'VærApp-Student/1.0 (student@skole.no)'}
    url = f"https://api.met.no/weatherapi/locationforecast/2.0/compact?lat={lat}&lon={lon}"
    
    try:
        response = requests.get(url, headers=headers, timeout=10)
        
        if response.status_code != 200:
            print(f"API-feil: Status {response.status_code}")
            return []
        
        data = response.json()
        timeseries = data['properties']['timeseries']
        
        forecast = []
        
        # Gå gjennom de første 'hours' prognosene
        for i, entry in enumerate(timeseries[:hours]):
            try:
                instant_data = entry['data']['instant']['details']
                
                forecast_entry = {
                    'tidspunkt': entry['time'],
                    'temperatur': instant_data['air_temperature'],
                    'luftfuktighet': instant_data['relative_humidity'],
                    'vindhastighet': instant_data['wind_speed']
                }
                
                forecast.append(forecast_entry)
                
            except KeyError:
                # Noen tidspunkter kan mangle data
                continue
        
        return forecast
        
    except Exception as e:
        print(f"Feil ved henting av prognose: {e}")
        return []

# Test funksjonen
oslo_forecast = get_hourly_forecast(59.9139, 10.7522, 8)

if oslo_forecast:
    print("📊 8-timers prognose for Oslo:")
    
    # Konverter til DataFrame for pen visning
    df = pd.DataFrame(oslo_forecast)
    
    # Forenkle tidspunkt-visning
    df['tid'] = pd.to_datetime(df['tidspunkt']).dt.strftime('%H:%M')
    
    # Vis de første prognosene
    print(df[['tid', 'temperatur', 'luftfuktighet', 'vindhastighet']].head(8).to_string(index=False))
else:
    print("❌ Kunne ikke hente værprognose")

📊 8-timers prognose for Oslo:
  tid  temperatur  luftfuktighet  vindhastighet
09:00         9.3           58.8            1.4
10:00        10.7           55.0            1.2
11:00        12.0           49.8            1.3
12:00        12.8           48.7            1.3
13:00        13.3           48.4            2.1
14:00        13.9           45.3            2.7
15:00        14.5           43.3            2.3
16:00        14.3           45.4            1.9


**Oppgave 6a:** Finn temperatur-ekstremene i prognosen

In [ ]:
# Analyser temperaturen i prognosen
if oslo_forecast:
    temperatures = [entry['temperatur'] for entry in oslo_forecast]
    
    min_temp = min(temperatures)
    max_temp = max(temperatures)
    avg_temp = sum(temperatures) / len(temperatures)
    
    print(f"🌡️ Temperaturanalyse for de neste {len(oslo_forecast)} timene:")
    print(f"Laveste: {min_temp}°C")
    print(f"Høyeste: {max_temp}°C")
    print(f"Gjennomsnitt: {avg_temp:.1f}°C")
    print(f"Temperaturspenn: {max_temp - min_temp}°C")

🌡️ Temperaturanalyse for de neste 8 timene:
Laveste: 9.3°C
Høyeste: 14.5°C
Gjennomsnitt: 12.6°C
Temperaturspenn: 5.199999999999999°C


## Oppgave 7: Sammenlign været i flere byer

La oss bruke funksjonene våre til å sammenligne været i flere norske byer.

In [ ]:
# Definer koordinater for norske byer
norske_byer = {
    'Oslo': (59.9139, 10.7522),
    'Bergen': (60.3913, 5.3221),
    'Trondheim': (63.4305, 10.3951),
    'Stavanger': (58.9700, 5.7331),
    'Tromsø': (69.6492, 18.9553)
}

# Hent værdata for alle byene
by_weather = []

print("🌍 Henter værdata for norske byer...")

for by_navn, (lat, lon) in norske_byer.items():
    print(f"  Henter data for {by_navn}...")
    
    weather = get_current_weather(lat, lon)
    
    if weather:
        weather['by'] = by_navn
        by_weather.append(weather)
    else:
        print(f"    ⚠️ Kunne ikke hente data for {by_navn}")

# Vis resultatene
if by_weather:
    print("\n🌤️ Værsammenligning:")
    
    # Konverter til DataFrame
    df = pd.DataFrame(by_weather)
    
    # Vis relevante kolonner
    weather_table = df[['by', 'temperatur_celsius', 'luftfuktighet_prosent', 'vindhastighet_ms']].round(1)
    weather_table.columns = ['By', 'Temperatur (°C)', 'Luftfuktighet (%)', 'Vind (m/s)']
    
    print(weather_table.to_string(index=False))
    
    # Finn ekstremene
    varmeste_idx = df['temperatur_celsius'].idxmax()
    kaldeste_idx = df['temperatur_celsius'].idxmin()
    
    print(f"\n🔥 Varmeste by: {df.loc[varmeste_idx, 'by']} ({df.loc[varmeste_idx, 'temperatur_celsius']}°C)")
    print(f"🧊 Kaldeste by: {df.loc[kaldeste_idx, 'by']} ({df.loc[kaldeste_idx, 'temperatur_celsius']}°C)")
else:
    print("❌ Kunne ikke hente værdata for noen byer")

🌍 Henter værdata for norske byer...
  Henter data for Oslo...
  Henter data for Bergen...
  Henter data for Trondheim...
  Henter data for Stavanger...
  Henter data for Tromsø...

🌤️ Værsammenligning:
       By  Temperatur (°C)  Luftfuktighet (%)  Vind (m/s)
     Oslo              9.3               58.8         1.4
   Bergen              6.9               82.0         0.8
Trondheim              6.2               88.7         2.2
Stavanger              8.2               76.1         2.8
   Tromsø              4.4               88.7         5.1

🔥 Varmeste by: Oslo (9.3°C)
🧊 Kaldeste by: Tromsø (4.4°C)


**Oppgave 7a:** Finn byen med høyest vindhastighet

In [ ]:
# Din kode her:
# Finn byen med høyest vindhastighet
if by_weather:
    # vindigste_idx = df['vindhastighet_ms'].idxmax()
    # print(f"💨 Vindigste by: {df.loc[vindigste_idx, 'by']} ({df.loc[vindigste_idx, 'vindhastighet_ms']} m/s)")
    pass

## Oppgave 8: Avansert - Daglig værsammendrag

La oss lage en funksjon som gir oss et sammendrag av været for de neste dagene.

In [ ]:
def get_daily_summary(lat, lon, days=3):
    """
    Henter daglig værsammendrag for de neste dagene.
    
    Args:
        lat (float): Breddegrad
        lon (float): Lengdegrad
        days (int): Antall dager
    
    Returns:
        DataFrame: Daglig værsammendrag
    """
    
    # Hent prognose for mange timer (24 timer * antall dager)
    forecast = get_hourly_forecast(lat, lon, hours=days * 24)
    
    if not forecast:
        return None
    
    # Konverter til DataFrame
    df = pd.DataFrame(forecast)
    
    # Legg til dato-kolonne
    df['datetime'] = pd.to_datetime(df['tidspunkt'])
    df['dato'] = df['datetime'].dt.date
    
    # Gruppér etter dag og beregn statistikk
    daily_summary = df.groupby('dato').agg({
        'temperatur': ['min', 'max', 'mean'],
        'luftfuktighet': 'mean',
        'vindhastighet': ['mean', 'max']
    }).round(1)

    # Forenkle kolonnenavn
    daily_summary.columns = [
        'Min_temp', 'Max_temp', 'Snitt_temp',
        'Snitt_luftfuktighet', 'Snitt_vind', 'Max_vind'
    ]

    return daily_summary

# Test funksjonen for Oslo
print("📅 3-dagers værsammendrag for Oslo:")
oslo_summary = get_daily_summary(59.9139, 10.7522, 3)

if oslo_summary is not None:
    print(oslo_summary)

    print(f"📊 Analyse:")
    print(f"Varmeste dag: {oslo_summary['Max_temp'].max()}°C")
    print(f"Kaldeste dag: {oslo_summary['Min_temp'].min()}°C")
    print(f"Største temperaturspenn på en dag: {(oslo_summary['Max_temp'] - oslo_summary['Min_temp']).max()}°C")
else:
    print("❌ Kunne ikke hente daglig sammendrag")


📅 3-dagers værsammendrag for Oslo:
            Min_temp  Max_temp  Snitt_temp  Snitt_luftfuktighet  Snitt_vind  \
dato                                                                          
2025-09-23       8.3      14.5        11.3                 59.1         1.8   
2025-09-24       6.3      15.9        10.6                 62.9         1.6   
2025-09-25       6.5      14.8         9.4                 75.2         1.3   
2025-09-26       7.6      15.3        11.0                 85.1         1.2   
2025-09-27       7.3      15.3        10.9                 84.3         1.4   
2025-09-28       7.8      14.9        10.9                 86.6         1.6   
2025-09-29       8.7      14.7        11.6                 89.0         1.6   
2025-09-30      10.0      14.5        11.5                 89.5         1.5   

            Max_vind  
dato                  
2025-09-23       2.7  
2025-09-24       2.1  
2025-09-25       2.4  
2025-09-26       1.4  
2025-09-27       1.8  
2025-09-28   

## Bonusoppgaver

Hvis du har kommet så langt, kan du prøve disse utfordringsoppgavene!

### Bonusoppgave 1: Lag en værvarsel-klasse

Organiser funksjonaliteten i en Python-klasse for bedre struktur.

In [ ]:
class WeatherAPI:
    """Klasse for å hente værdata fra Yr.no API"""
    
    def __init__(self, user_agent="VærApp-Student/1.0 (student@skole.no)"):
        self.base_url = "https://api.met.no/weatherapi/locationforecast/2.0/compact"
        self.headers = {'User-Agent': user_agent}
    
    def _make_request(self, lat, lon):
        """Hjelpemetode for å gjøre API-forespørsel"""
        url = f"{self.base_url}?lat={lat}&lon={lon}"
        
        try:
            response = requests.get(url, headers=self.headers, timeout=10)
            response.raise_for_status()  # Kaster exception ved HTTP-feil
            return response.json()
        except requests.exceptions.RequestException as e:
            print(f"API-feil: {e}")
            return None
    
    def get_current_weather(self, lat, lon):
        """Hent nåværende vær"""
        # Implementer denne metoden
        # Din kode her:
        pass
    
    def get_forecast(self, lat, lon, hours=12):
        """Hent timesprognose"""
        # Implementer denne metoden
        # Din kode her:
        pass

# Test klassen din
# weather_api = WeatherAPI()
# weather = weather_api.get_current_weather(59.9139, 10.7522)
# print(weather)

### Bonusoppgave 2: Robust feilhåndtering

Implementer en funksjon som prøver på nytt hvis API-et ikke svarer.

In [ ]:
import time

def robust_weather_request(lat, lon, max_retries=3, delay=1):
    """
    Robust API-forespørsel med retry-logikk.
    
    Args:
        lat, lon: Koordinater
        max_retries: Maksimalt antall forsøk
        delay: Sekunder å vente mellom forsøk
    """
    
    for attempt in range(max_retries):
        try:
            print(f"Forsøk {attempt + 1}/{max_retries}...")
            
            # Din kode her:
            # Implementer logikk som:
            # 1. Prøver å hente data
            # 2. Returnerer data hvis vellykket
            # 3. Venter 'delay' sekunder og prøver igjen ved feil
            # 4. Gir opp etter max_retries forsøk
            
            pass
            
        except Exception as e:
            print(f"Forsøk {attempt + 1} feilet: {e}")
            
            if attempt < max_retries - 1:
                print(f"Venter {delay} sekunder før nytt forsøk...")
                time.sleep(delay)
            else:
                print("Alle forsøk feilet!")
                return None

# Test funksjonen
# robust_data = robust_weather_request(59.9139, 10.7522)
# print(robust_data)

### Bonusoppgave 3: Værvarsel med nedbør

Yr.no API-et inneholder også informasjon om nedbør. Utforsk datastrukturen og hent ut nedbørsprognoser.

In [ ]:
# Utforsk 'next_1_hours' og 'next_6_hours' data
# Disse inneholder nedbørsprognoser

def get_precipitation_forecast(lat, lon):
    """Hent nedbørsprognose for de neste timene"""
    
    # Din kode her:
    # Tips: Se på 'next_1_hours' og 'next_6_hours' i datastrukturen
    # Disse inneholder 'precipitation_amount' og 'precipitation_amount_max'
    
    pass

# Test funksjonen
# precipitation = get_precipitation_forecast(59.9139, 10.7522)
# print(precipitation)

## Refleksjonsspørsmål

Ta deg tid til å tenke over det du har lært:

1. **API-forståelse**: Hva er forskjellen mellom et API og en vanlig nettside?

2. **Feilhåndtering**: Hvorfor er det viktig å håndtere feil når man arbeider med API-er?

3. **Datastruktur**: Yr.no returnerer mye data. Hvordan bestemte du hvilke verdier som var viktigst å hente ut?

4. **Etikk**: Hvilke etiske hensyn bør man ta når man bruker andres API-er?

5. **Utvidelser**: Hvilke andre funksjoner kunne du tenke deg å legge til i en værapp?

Skriv ned tankene dine i cellen under:

**Mine refleksjoner:**

<!-- Skriv dine tanker her -->



## Oppsummering

🎉 **Gratulerer! Du har nå lært:**

✅ **REST API-konsepter**
- Hva et API er og hvordan det fungerer
- HTTP-metoder (GET, POST, etc.)
- Status koder (200, 404, etc.)

✅ **requests-biblioteket**
- Gjøre HTTP-forespørsler
- Håndtere JSON-data
- Sette headers og parametere

✅ **Yr.no API**
- Hente værdata for spesifikke koordinater
- Forstå komplekse datastrukturer
- Hente både nåværende vær og prognoser

✅ **Databehandling**
- Pakke ut relevant informasjon fra API-svar
- Sammenligne data fra flere kilder
- Lage sammendrag og statistikk

✅ **Feilhåndtering**
- Try/except blokker
- Håndtere nettverksfeil
- Validere data

### Neste steg i læringen:

🔗 **Utforsk andre API-er:**
- [JSONPlaceholder](https://jsonplaceholder.typicode.com/) - Test-API
- [OpenWeatherMap](https://openweathermap.org/api) - Alternativt vær-API
- [REST Countries](https://restcountries.com/) - Landeinformasjon

📚 **Lær mer om:**
- API-autentisering (API-nøkler, OAuth)
- GraphQL (alternativ til REST)
- Rate limiting og caching
- Asynkron programmering med `asyncio`

🛠️ **Prosjektideer:**
- Bygg en værapp med grafisk brukergrensesnitt
- Lag en bot som sender daglige værmeldinger
- Kombiner flere API-er (vær + kart + nyheter)
- Lag en værlogger som lagrer data i en database

### Viktige ressurser:

- 📖 [Requests dokumentasjon](https://docs.python-requests.org/)
- 🌤️ [Yr.no API dokumentasjon](https://api.met.no/weatherapi/)
- 🔧 [HTTP statuskoder](https://httpstatuses.com/)
- 📊 [Pandas dokumentasjon](https://pandas.pydata.org/docs/)

**Lykke til med videre utforskning av API-verdenen! 🚀**